### PR TITLE
build: configure build with pyproject.toml

### PR DESCRIPTION
**Pull request by jsstevenson**
I am not sure I can test that this produces a totally valid release, because `setuptools_scm` will [render a version number](https://setuptools-scm.readthedocs.io/en/latest/usage/#default-versioning-scheme) that is incompatible with PyPI if you do so from a commit that doesn't have a version tag on it.  I'm pretty sure it's correct, though.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    "pylint",
+    "yapf",
+    "smart_open",
+    # docs
+    "sphinx",
+    "sphinx_rtd_theme",
+    "restview",
+    # idk?  TODO figure it out
+    "ipython",
+    "jedi<0.18",
```
Comment (jsstevenson) [inline pyproject.toml:75]: truly not sure what these are doing here

File: setup.cfg [hunk 2]
```diff
@@ -1,127 +0,0 @@
...
-max-complexity = 10
-ignore = E129,E221,E241,E251,E303,W291
-
-[tool:pytest]
-# TODO: add src to testpaths to trigger doctests. However, this
-# requires injecting fixtures, which requires conftest.py in src/,
-# which requires packaging changes. Deferred.
-testpaths = src tests
-doctest_optionflags = ALLOW_UNICODE ALLOW_BYTES ELLIPSIS IGNORE_EXCEPTION_DETAIL NORMALIZE_WHITESPACE
-addopts = --cov-config=setup.cfg --cov-report=term-missing --cov=ga4gh
```
Comment (jsstevenson) [inline setup.cfg:127]: `setup.cfg` is given as the pytest-cov config target but no actual configuration options are set. I removed it.

File: setup.cfg [hunk 3]
```diff
@@ -1,127 +0,0 @@
...
-
-[build_sphinx]
-all_files  = 1
-
-# http://pep8.readthedocs.org/en/latest/intro.html#error-codes
-[flake8]
-max-line-length = 120
-exclude = tests/*
-max-complexity = 10
-ignore = E129,E221,E241,E251,E303,W291
```
Comment (jsstevenson) [inline setup.cfg:119]: Didn't add as a separate `.flake8` file because flake8 isn't ever installed as part of dev dependencies

File: setup.cfg [hunk 4]
```diff
@@ -1,127 +0,0 @@
...
-    ipython
-    jupyter
-    tabulate
-
-
-[aliases]
-test = pytest
-
-[bdist_wheel]
-universal = 1
```
Comment (jsstevenson) [inline setup.cfg:109]: This is for retaining compatibility with Python 2 (https://wheel.readthedocs.io/en/stable/user_guide.html#building-wheels) -- we can remove it

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,109 @@
+[project]
+name = "ga4gh.vrs"
+authors = [
+    {name = "Larry Babb", email = "lbabb@broadinstitute.org"},
+    {name = "Reece Hart", email = "reecehart@gmail.com"},
+    {name = "Andreas Prlic", email = "andreas.prlic@gmail.com"},
+    {name = "Alex Wagner", email = "alex.wagner@nationwidechildrens.org"},
```
Comment (jsstevenson) [inline pyproject.toml:7]: previously this was just Reece, but I went with all of the VR heads for the sake of consistency

File: pyproject.toml (hunk 1 above)
Comment (theferrit32) [reply pyproject.toml:75]: @jsstevenson I would delete these, I don't know what jedi is used for, I don't think it's necessary. `ipython` is only needed for notebooks

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    # docs
+    "sphinx",
+    "sphinx_rtd_theme",
+    "restview",
+    # idk?  TODO figure it out
+    "ipython",
+    "jedi<0.18",
+]
+notebooks = [
+    "ipython",
```
Comment (theferrit32) [inline pyproject.toml:78]: I would also delete `ipython` from here, it's pulled in by `jupyter` already

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    "vcrpy",
+    "pyyaml",
+    # style
+    "pylint",
+    "yapf",
+    # docs
+    "sphinx",
+    "sphinx_rtd_theme",
+    "restview",
+    "smart_open",  # TODO remove?
```
Comment (jsstevenson) [inline pyproject.toml:72]: another one -- anybody (@theferrit32, others?) have a clue if this is needed? not finding anything from a quick grep

Comment (theferrit32) [reply pyproject.toml:72]: `smart_open` had been used by Reece in some scripts in some biocommons projects but is not used here, you can remove

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: Apache Software License",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
```
Comment (korikuzma) [inline pyproject.toml:25]: Note to update GH actions to also test 3.12

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    "pydantic~=2.1",
+]
+
+[project.optional-dependencies]
+extras = [
+    "psycopg2-binary",
+    "biocommons.seqrepo>=0.5.1",
+    "bioutils>=0.5.2",
+    "hgvs>=1.4",
+    "requests",
```
Comment (korikuzma) [inline pyproject.toml:54]: Should `requests` be in `dependencies` since it's used in dataproxy module (not in extras)?

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    "variation"
+]
+requires-python = ">=3.8"
+dynamic = ["version"]
+dependencies = [
+    "bioutils",
+    "canonicaljson",
+    "coloredlogs",
+    "numpy",
+    "pyyaml",
```
Comment (korikuzma) [inline pyproject.toml:44]: I think we can remove these. I don't see canonicaljson, coloredlogs, or numpy being used anywhere and I think pyyaml belongs in dev and notebooks

```suggestion
```

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    "yapf",
+    # docs
+    "sphinx",
+    "sphinx_rtd_theme",
+    "restview",
+    "smart_open",  # TODO remove?
+]
+notebooks = [
+    "jupyter",
+    "tabulate",
```
Comment (korikuzma) [inline pyproject.toml:76]:
```suggestion
    "tabulate",
    "pyyaml",
```

**Review by korikuzma — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 10 above)
Comment (jsstevenson) [reply pyproject.toml:44]: Yeah looks like a lot of this was removed in 64fee4c1a441e1623cda18734aa28c9137afb14b

Comment (jsstevenson) [reply pyproject.toml:44]: also, numpy????

**Review by korikuzma — APPROVED** [review]

**Pull request merged**